[MLIR][build] Build fails with -DCMAKE_BUILD_TYPE=Debug -DBUILD_SHARED_LIBS=1

```
$ cmake -S llvm -B build-shlib -G Ninja \
    -DCMAKE_EXPORT_COMPILE_COMMANDS=1 \
    -DLLVM_ENABLE_PROJECTS="mlir" \
    -DCMAKE_BUILD_TYPE=Debug \
    -DBUILD_SHARED_LIBS=1 \
    -DLLVM_TARGETS_TO_BUILD="X86" \
    -DLLVM_ENABLE_BINDINGS=0 \
    -DLLVM_CCACHE_BUILD=1

-- The C compiler identification is GNU 13.3.0                                                                                                                                                                                                          
-- The CXX compiler identification is GNU 13.3.0                                                                                                                                                                                                        
-- The ASM compiler identification is GNU                                                                                                                                                                                                               
-- Found assembler: /usr/bin/cc                                                                                                                                                                                                                         
-- Detecting C compiler ABI info                                                                                                                                                                                                                        
-- Detecting C compiler ABI info - done                                                                                                                                                                                                                 
-- Check for working C compiler: /usr/bin/cc - skipped                                                                                                                                                                                                  
-- Detecting C compile features                                                                                                                                                                                                                         
-- Detecting C compile features - done                                                                                                                                                                                                                  
-- Detecting CXX compiler ABI info                                                                                                                                                                                                                      
-- Detecting CXX compiler ABI info - done                                                                                                                                                                                                               
-- Check for working CXX compiler: /usr/bin/c++ - skipped                                                                                                                                                                                               
-- Detecting CXX compile features                                                                                           
-- Detecting CXX compile features - done                                                                                                                                                                                                                
-- bolt project is disabled                                                                                                                                                                                                                             
-- clang project is disabled                                                                                                                                                                                                                            
-- clang-tools-extra project is disabled                                                                                                                                                                                                                
-- cross-project-tests project is disabled                                                                                                                                                                                                              
-- lld project is disabled                                                                                                                                                                                                                              
-- lldb project is disabled                                                                                                                                                                                                                             
-- mlir project is enabled                                                                                                                                                                                                                              
-- polly project is disabled                                                                                                                                                                                                                            
-- flang project is disabled                                                                                                                                                                                                                            
-- libc project is disabled                                                                                                                                                                                                                             
-- compiler-rt project is disabled                                                                                                                                                                                                                      
Using ccache with precompiled headers with non-Clang compilers is not supported. CMAKE_DISABLE_PRECOMPILE_HEADERS will be set to ON. Pass -DCMAKE_DISABLE_PRECOMPILE_HEADERS=OFF to override this.
-- Doxygen disabled.                                          
-- Found Python3: /usr/bin/python3 (found suitable version "3.12.3", minimum required is "3.8") found components: Interpreter 
-- Looking for __GLIBC__                                      
-- Looking for __GLIBC__ - found                              
-- Looking for valgrind/valgrind.h                            
-- Looking for valgrind/valgrind.h - not found                
-- Looking for FE_ALL_EXCEPT                                  
-- Looking for FE_ALL_EXCEPT - found                          
-- Looking for FE_INEXACT                                     
-- Looking for FE_INEXACT - found                             
-- Performing Test HAVE_BUILTIN_THREAD_POINTER                
-- Performing Test HAVE_BUILTIN_THREAD_POINTER - Success      
-- Looking for CrashReporterClient.h                          
-- Looking for CrashReporterClient.h - not found              
-- Looking for linux/magic.h                                  
-- Looking for linux/magic.h - found                          
-- Looking for pthread_create in pthread                      
-- Looking for pthread_create in pthread - found              
-- Looking for pthread_rwlock_init in pthread                 
-- Looking for pthread_rwlock_init in pthread - found         
-- Looking for pthread_mutex_lock in pthread                  
-- Looking for pthread_mutex_lock in pthread - found          
-- Looking for dlopen in dl                                   
-- Looking for dlopen in dl - found                           
-- Looking for shm_open in rt                                 
-- Looking for shm_open in rt - found                         
-- Looking for pfm_initialize in pfm                          
-- Looking for pfm_initialize in pfm - not found              
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD                    
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE                                        
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.3")                                                      
-- Looking for compress2                                      
-- Looking for compress2 - found                              
-- Found LibXml2: /usr/lib/x86_64-linux-gnu/libxml2.so (found version "2.9.14")                                             
-- Looking for xmlReadMemory                                  
-- Looking for xmlReadMemory - found                          
-- Could NOT find LibEdit (missing: LibEdit_INCLUDE_DIRS LibEdit_LIBRARIES)                                                 
-- Looking for arc4random                                     
-- Looking for arc4random - found                             
-- Looking for backtrace                                      
-- Looking for backtrace - found                              
-- backtrace facility detected in default set of libraries    
-- Found Backtrace: /usr/include                              
-- Performing Test C_SUPPORTS_WERROR_UNGUARDED_AVAILABILITY_NEW                                                             
-- Performing Test C_SUPPORTS_WERROR_UNGUARDED_AVAILABILITY_NEW - Failed                                                    
-- Looking for __register_frame                               
-- Looking for __register_frame - found                       
-- Looking for __deregister_frame                             
-- Looking for __deregister_frame - found                     
-- Looking for __unw_add_dynamic_fde                          
-- Looking for __unw_add_dynamic_fde - not found              
-- Looking for _Unwind_Backtrace                              
-- Looking for _Unwind_Backtrace - found                      
-- Looking for getpagesize                                    
-- Looking for getpagesize - found                            
-- Looking for sysconf                                        
-- Looking for sysconf - found                                
-- Looking for getrusage                                      
-- Looking for getrusage - found                              
-- Looking for isatty                                         
-- Looking for isatty - found                                 
-- Looking for futimens                                       
-- Looking for futimens - found                               
-- Looking for futimes                                        
-- Looking for futimes - found                                
-- Looking for getauxval                                      
-- Looking for getauxval - found                              
-- Looking for sigaltstack                                    
-- Looking for sigaltstack - found                            
-- Looking for mallctl                                        
-- Looking for mallctl - not found                            
-- Looking for mallinfo                                       
-- Looking for mallinfo - found                               
-- Looking for mallinfo2                                      
-- Looking for mallinfo2 - found                              
-- Looking for malloc_zone_statistics                         
-- Looking for malloc_zone_statistics - not found             
-- Looking for posix_spawn                                    
-- Looking for posix_spawn - found                            
-- Looking for pread                                          
-- Looking for pread - found                                  
-- Looking for sbrk                                           
-- Looking for sbrk - found                                   
-- Looking for strerror_r                                     
-- Looking for strerror_r - found                             
-- Looking for strerror_s                                     
-- Looking for strerror_s - not found                         
-- Looking for setenv 
-- Looking for setenv - found                                 
-- Performing Test HAVE_STRUCT_STAT_ST_MTIMESPEC_TV_NSEC      
-- Performing Test HAVE_STRUCT_STAT_ST_MTIMESPEC_TV_NSEC - Failed                                                           
-- Performing Test HAVE_STRUCT_STAT_ST_MTIM_TV_NSEC           
-- Performing Test HAVE_STRUCT_STAT_ST_MTIM_TV_NSEC - Success                                                               
-- Looking for pthread_getname_np                             
-- Looking for pthread_getname_np - found                     
-- Looking for pthread_setname_np                             
-- Looking for pthread_setname_np - found                     
-- Looking for pthread_get_name_np                            
-- Looking for pthread_get_name_np - not found                
-- Looking for pthread_set_name_np                            
-- Looking for pthread_set_name_np - not found                
-- Looking for dlopen                                         
-- Looking for dlopen - found                                 
-- Looking for proc_pid_rusage                                
-- Looking for proc_pid_rusage - not found                    
-- Performing Test HAVE_CXX_ATOMICS_WITHOUT_LIB               
-- Performing Test HAVE_CXX_ATOMICS_WITHOUT_LIB - Success     
-- Performing Test HAVE_CXX_ATOMICS64_WITHOUT_LIB             
-- Performing Test HAVE_CXX_ATOMICS64_WITHOUT_LIB - Success   
-- Performing Test LLVM_HAS_ATOMICS                           
-- Performing Test LLVM_HAS_ATOMICS - Success                 
-- LLVM host triple: x86_64-unknown-linux-gnu                 
-- Native target architecture is X86                          
-- Threads enabled.                                           
-- Ninja version: 1.11.1                                      
-- Found OCaml: /home/drank/.opam/default/bin/ocamlfind       
-- OCaml bindings disabled.                                   
-- Found Python module pygments                               
-- Found Python module pygments.lexers.c_cpp                  
-- Found Python module yaml                                   
-- LLVM default target triple: x86_64-unknown-linux-gnu       
-- Performing Test SUPPORTS_LIBCXX_HARDENING_MODE             
-- Performing Test SUPPORTS_LIBCXX_HARDENING_MODE - Success   
-- Performing Test LLVM_USES_LIBSTDCXX                        
-- Performing Test LLVM_USES_LIBSTDCXX - Success              
-- Performing Test LLVM_DEFAULT_TO_GLIBCXX_USE_CXX11_ABI      
-- Performing Test LLVM_DEFAULT_TO_GLIBCXX_USE_CXX11_ABI - Success                                                          
-- Performing Test C_SUPPORTS_FPIC                            
-- Performing Test C_SUPPORTS_FPIC - Success                  
-- Performing Test CXX_SUPPORTS_FPIC                          
-- Performing Test CXX_SUPPORTS_FPIC - Success                
-- Building with -fPIC                                        
-- Performing Test C_SUPPORTS_FNO_SEMANTIC_INTERPOSITION      
-- Performing Test C_SUPPORTS_FNO_SEMANTIC_INTERPOSITION - Success                                                          
-- Performing Test CXX_SUPPORTS_FNO_SEMANTIC_INTERPOSITION    
-- Performing Test CXX_SUPPORTS_FNO_SEMANTIC_INTERPOSITION - Success                                                        
-- Performing Test SUPPORTS_FVISIBILITY_INLINES_HIDDEN_FLAG   
-- Performing Test SUPPORTS_FVISIBILITY_INLINES_HIDDEN_FLAG - Success                                                       
-- Performing Test C_SUPPORTS_CXX98_COMPAT_EXTRA_SEMI_FLAG    
-- Performing Test C_SUPPORTS_CXX98_COMPAT_EXTRA_SEMI_FLAG - Failed                                                         
-- Performing Test CXX_SUPPORTS_CXX98_COMPAT_EXTRA_SEMI_FLAG  
-- Performing Test CXX_SUPPORTS_CXX98_COMPAT_EXTRA_SEMI_FLAG - Failed                                                       
-- Performing Test CXX_SUPPORTS_SUGGEST_OVERRIDE_FLAG
-- Performing Test CXX_SUPPORTS_SUGGEST_OVERRIDE_FLAG - Success                                                                                                                                                         (18 results) 18:55:50 [843/3372]
-- Performing Test CXX_WSUGGEST_OVERRIDE_ALLOWS_ONLY_FINAL    
-- Performing Test CXX_WSUGGEST_OVERRIDE_ALLOWS_ONLY_FINAL - Success                                                        
-- Performing Test C_WCOMMENT_ALLOWS_LINE_WRAP                
-- Performing Test C_WCOMMENT_ALLOWS_LINE_WRAP - Failed       
-- Performing Test C_SUPPORTS_CTAD_MAYBE_UNSPPORTED_FLAG      
-- Performing Test C_SUPPORTS_CTAD_MAYBE_UNSPPORTED_FLAG - Failed                                                           
-- Performing Test CXX_SUPPORTS_CTAD_MAYBE_UNSPPORTED_FLAG    
-- Performing Test CXX_SUPPORTS_CTAD_MAYBE_UNSPPORTED_FLAG - Success                                                        
-- Performing Test LINKER_SUPPORTS_COLOR_DIAGNOSTICS          
-- Performing Test LINKER_SUPPORTS_COLOR_DIAGNOSTICS - Failed                                                               
-- Precompiled headers disabled.                              
-- Looking for os_signpost_interval_begin                     
-- Looking for os_signpost_interval_begin - not found         
-- Looking for flock                                          
-- Looking for flock - found                                  
-- Linker detection: GNU ld                                   
-- Performing Test LLVM_LINKER_SUPPORTS_B_SYMBOLIC_FUNCTIONS  
-- Performing Test LLVM_LINKER_SUPPORTS_B_SYMBOLIC_FUNCTIONS - Success                                                      
-- Performing Test HAS_WERROR_GLOBAL_CTORS                    
-- Performing Test HAS_WERROR_GLOBAL_CTORS - Failed           
-- Looking for __x86_64__                                     
-- Looking for __x86_64__ - found                             
-- Found Git: /usr/bin/git (found version "2.43.0")           
-- Looking for logf128                                        
-- Looking for logf128 - found                                
-- Targeting X86                                              
-- Performing Test C_SUPPORTS_WERROR_IMPLICIT_FUNCTION_DECLARATION                                                          
-- Performing Test C_SUPPORTS_WERROR_IMPLICIT_FUNCTION_DECLARATION - Success                                                
-- Performing Test C_SUPPORTS_WUNDEF                          
-- Performing Test C_SUPPORTS_WUNDEF - Success                
-- Performing Test C_SUPPORTS_WERROR_MISMATCHED_TAGS          
-- Performing Test C_SUPPORTS_WERROR_MISMATCHED_TAGS - Failed                                                               
-- Performing Test CXX_SUPPORTS_WNO_UNUSED_BUT_SET_PARAMETER  
-- Performing Test CXX_SUPPORTS_WNO_UNUSED_BUT_SET_PARAMETER - Success                                                      
-- Performing Test C_SUPPORTS_WERROR_GLOBAL_CONSTRUCTOR       
-- Performing Test C_SUPPORTS_WERROR_GLOBAL_CONSTRUCTOR - Failed                                                            
-- Performing Test CXX_SUPPORTS_WERROR_GLOBAL_CONSTRUCTOR     
-- Performing Test CXX_SUPPORTS_WERROR_GLOBAL_CONSTRUCTOR - Failed                                                          
-- Performing Test COMPILER_SUPPORTS_WARNING_WEAK_VTABLES     
-- Performing Test COMPILER_SUPPORTS_WARNING_WEAK_VTABLES - Failed                                                          
-- Registering ExampleIRTransforms as a pass plugin (static build: OFF)                                                     
-- Registering Bye as a pass plugin (static build: OFF)       
-- Failed to find LLVM FileCheck                              
-- Google Benchmark version: v0.0.0, normalized to 0.0.0      
-- Looking for shm_open in rt                                 
-- Looking for shm_open in rt - found                         
-- Performing Test HAVE_CXX_FLAG_WALL                         
-- Performing Test HAVE_CXX_FLAG_WALL - Success               
-- Performing Test HAVE_CXX_FLAG_WEXTRA                       
-- Performing Test HAVE_CXX_FLAG_WEXTRA - Success             
-- Performing Test HAVE_CXX_FLAG_WSHADOW                      
-- Performing Test HAVE_CXX_FLAG_WSHADOW - Success            
-- Performing Test HAVE_CXX_FLAG_WFLOAT_EQUAL                 
-- Performing Test HAVE_CXX_FLAG_WFLOAT_EQUAL - Success       
-- Performing Test HAVE_CXX_FLAG_WOLD_STYLE_CAST              
-- Performing Test HAVE_CXX_FLAG_WOLD_STYLE_CAST - Success    
-- Performing Test HAVE_CXX_FLAG_WSUGGEST_OVERRIDE            
-- Performing Test HAVE_CXX_FLAG_WSUGGEST_OVERRIDE - Success  
-- Performing Test HAVE_CXX_FLAG_PEDANTIC
-- Performing Test HAVE_CXX_FLAG_PEDANTIC - Success           
-- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS              
-- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS - Success    
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32            
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32 - Failed   
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING             
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING - Success   
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED_DECLARATIONS  
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED_DECLARATIONS - Success                                                      
-- Performing Test HAVE_CXX_FLAG_FNO_EXCEPTIONS               
-- Performing Test HAVE_CXX_FLAG_FNO_EXCEPTIONS - Success     
-- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING             
-- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING - Success   
-- Performing Test HAVE_CXX_FLAG_WD654                        
-- Performing Test HAVE_CXX_FLAG_WD654 - Failed               
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY               
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY - Failed      
-- Performing Test HAVE_CXX_FLAG_COVERAGE                     
-- Performing Test HAVE_CXX_FLAG_COVERAGE - Success           
-- Compiling and running to test HAVE_GNU_POSIX_REGEX         
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile  
-- Compiling and running to test HAVE_POSIX_REGEX             
-- Performing Test HAVE_POSIX_REGEX -- success                
-- Compiling and running to test HAVE_STEADY_CLOCK            
-- Performing Test HAVE_STEADY_CLOCK -- success               
-- Compiling and running to test HAVE_PTHREAD_AFFINITY        
-- Performing Test HAVE_PTHREAD_AFFINITY -- success           
-- Configuring done (12.4s)                                   
-- Generating done (2.4s)                                     
-- Build files have been written to: /home/drank/dev/compilers/llvm-project/build-shlib                                     

$ ninja -C build-shlib -j12

[4128/5933] Linking CXX shared library lib/libMLIROpenMPDialect.so.23.0git                                                  
FAILED: lib/libMLIROpenMPDialect.so.23.0git                   
: && /usr/bin/c++ -fPIC -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough 
-Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-dangling-reference -Wno-redundant-move -Wno-pessimizing-move -Wno-array-bounds -Wno-stringop-overread -Wno-dangling-pointer -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -W
no-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -Wundef -Wno-unused-but-set-parameter -g  -Wl,-z,defs -Wl,-z,nodelete   -Wl,-rpath-link,/home/drank/dev/compilers/llvm-project/build-shlib/./lib -shared -Wl,-soname
,libMLIROpenMPDialect.so.23.0git -o lib/libMLIROpenMPDialect.so.23.0git tools/mlir/lib/Dialect/OpenMP/CMakeFiles/obj.MLIROpenMPDialect.dir/IR/OpenMPDialect.cpp.o  -Wl,-rpath,"\$ORIGIN/../lib:/home/drank/dev/compilers/llvm-project/build-shlib/lib:" 
 lib/libMLIRLLVMDialect.so.23.0git  lib/libMLIRFuncDialect.so.23.0git  lib/libMLIROpenACCMPCommon.so.23.0git  lib/libMLIRDataLayoutInterfaces.so.23.0git  lib/libMLIRInferIntRangeInterface.so.23.0git  lib/libMLIRMemorySlotInterfaces.so.23.0git  lib/
libMLIRPtrMemorySpaceInterfaces.so.23.0git  lib/libLLVMCore.so.23.0git  lib/libLLVMBinaryFormat.so.23.0git  lib/libMLIRControlFlowInterfaces.so.23.0git  lib/libMLIRFunctionInterfaces.so.23.0git  lib/libMLIRCallInterfaces.so.23.0git  lib/libMLIRInfe
rTypeOpInterface.so.23.0git  lib/libMLIRSideEffectInterfaces.so.23.0git  lib/libMLIRIR.so.23.0git  lib/libMLIRSupport.so.23.0git  lib/libLLVMSupport.so.23.0git  -Wl,-rpath-link,/home/drank/dev/compilers/llvm-project/build-shlib/lib && :
/usr/bin/ld: tools/mlir/lib/Dialect/OpenMP/CMakeFiles/obj.MLIROpenMPDialect.dir/IR/OpenMPDialect.cpp.o: in function `llvm::Triple::normalize[abi:cxx11](llvm::Triple::CanonicalForm) const':
/home/drank/dev/compilers/llvm-project/llvm/include/llvm/TargetParser/Triple.h:412:(.text._ZNK4llvm6Triple9normalizeB5cxx11ENS0_13CanonicalFormE[_ZNK4llvm6Triple9normalizeB5cxx11ENS0_13CanonicalFormE]+0x4c): undefined reference to `llvm::Triple::no
rmalize[abi:cxx11](llvm::StringRef, llvm::Triple::CanonicalForm)'                                                           
collect2: error: ld returned 1 exit status                    
[4139/5933] Building CXX object tools/mlir/lib/Dialect/Spars...bj.MLIRSparseTensorPipelines.dir/SparseTensorPipelines.cpp.o 
ninja: build stopped: subcommand failed.
```

--- a/mlir/lib/Dialect/LLVMIR/CMakeLists.txt
+++ b/mlir/lib/Dialect/LLVMIR/CMakeLists.txt
@@ -60,11 +60,13 @@ add_mlir_dialect_library(MLIRNVVMDialect
   LINK_COMPONENTS
   AsmParser
   Core
+  TargetParser
 
   LINK_LIBS PUBLIC
   MLIRIR
   MLIRLLVMDialect
   MLIRGPUDialect
+  MLIROpenMPDialect
   MLIRSideEffectInterfaces
   MLIRInferIntRangeInterface
   )
@@ -128,10 +130,12 @@ add_mlir_dialect_library(MLIRXeVMDialect
   LINK_COMPONENTS
   AsmParser
   Core
+  TargetParser
 
   LINK_LIBS PUBLIC
   MLIRDialectUtils
   MLIRIR
   MLIRLLVMDialect
+  MLIROpenMPDialect
   MLIRSideEffectInterfaces
 )

--- a/mlir/lib/Dialect/OpenMP/CMakeLists.txt
+++ b/mlir/lib/Dialect/OpenMP/CMakeLists.txt
@@ -12,6 +12,9 @@ add_mlir_dialect_library(MLIROpenMPDialect
   MLIROpenMPOpsInterfacesIncGen
   MLIROpenMPTypeInterfacesIncGen
 
+  LINK_COMPONENTS
+  TargetParser
+
   LINK_LIBS PUBLIC
   MLIRIR
   MLIRLLVMDialect

